std/json.zig

```diff
@@ -0,0 +1,1308 @@
+                }
+                std.debug.warn("}}");
+            },
+        }
+    }
+
+    pub fn toStringIndent(self: &const Value, indent: usize) void {
+        if (indent == 0) {
+            self.toString();
+        } else {
+            self.toStringIndentLevel(indent, 0);
+        }
+    }
+
+    fn toStringIndentLevel(self: &const Value, indent: usize, level: usize) void {
+        switch (*self) {
+            Value.Null => {
+                std.debug.warn("null");
+            },
+            Value.Bool => |inner| {
+                std.debug.warn("{}", inner);
+            },
+            Value.Integer => |inner| {
+                std.debug.warn("{}", inner);
+            },
+            Value.Float => |inner| {
+                std.debug.warn("{.5}", inner);
+            },
+            Value.String => |inner| {
+                std.debug.warn("\"{}\"", inner);
+            },
+            Value.Array => |inner| {
+                var not_first = false;
+                std.debug.warn("[\n");
+
+                for (inner.toSliceConst()) |value| {
+                    if (not_first) {
+                        std.debug.warn(",\n");
+                    }
+                    not_first = true;
+                    padSpace(level + indent);
+                    value.toStringIndentLevel(indent, level + indent);
+                }
+                std.debug.warn("\n");
+                padSpace(level);
+                std.debug.warn("]");
+            },
+            Value.Object => |inner| {
+                var not_first = false;
+                std.debug.warn("{{\n");
+                var it = inner.iterator();
+
+                while (it.next()) |entry| {
+                    if (not_first) {
+                        std.debug.warn(",\n");
+                    }
+                    not_first = true;
+                    padSpace(level + indent);
+                    std.debug.warn("\"{}\": ", entry.key);
+                    entry.value.toStringIndentLevel(indent, level + indent);
+                }
+                std.debug.warn("\n");
+                padSpace(level);
+                std.debug.warn("}}");
+            },
+        }
+    }
+
+    fn padSpace(indent: usize) void {
+        var i: usize = 0;
+        while (i < indent) : (i += 1) {
+            std.debug.warn(" ");
+        }
+    }
+};
+
+// A non-stream JSON parser which constructs a tree of Value's.
+const JsonParser = struct {
+    allocator: &Allocator,
+    state: State,
+    copy_strings: bool,
+    // Stores parent nodes and un-combined Values.
+    // Worst case scenario we have nested key, values and so need two times the stack size.
+    stack: [2 * StreamingJsonParser.max_stack_size]Value,
+    stack_used: u16,
+
+    const State = enum {
+        ObjectKey,
+        ObjectValue,
+        ArrayValue,
+        Simple,
+    };
+
+    pub fn init(allocator: &Allocator, copy_strings: bool) JsonParser {
```

Comment:
What happens here if we pass an allocator derived from an ArenaAllocator and the use that as an input to our own ArenaAllocator? Are there other wrapping allocators which don't handle this correctly?

Reply:
The worst that can happen is inefficient space usage. Allocators that wrap other allocators respect the API that Allocator provides and so everything will work correctly.